### PR TITLE
Asymmetric density

### PR DESCRIPTION
**Pull request by bberenz**
Adds support for asymmetric densities and named density values.

### CODE REVIEW

File: src/qz/printer/info/NativePrinter.java [hunk 1]
```diff
@@ -152,22 +151,16 @@ public void setResolution(PrinterResolution resolution) {
         return resolution;
     }
 
-    public List<Integer> getResolutions() {
+    public List<PrinterResolution> getResolutions() {
         // TODO: Test/Implement supported resolutions for CUPS
```
Comment (tresf) [inline src/qz/printer/info/NativePrinter.java]: Unfortunately, due to this `TODO:` item and lack of density reporting from CUPS, we can't go higher than what's reported in the `.ppd` file as `*DefaultResolution: 300dpi` on my Dymo LabelWriter 450 Turbo on Mac.

The Dymo `.ppd` comes back with the following when parsing for `dpi`:

```diff
+ *DefaultResolution: 300dpi
  *Resolution 300dpi/300 DPI: "<</HWResolution[300 300]>>setpagedevice"
  *Resolution 300x600dpi/300x600 DPI: "<</HWResolution[300 600]>>setpagedevice"
  *UIConstraints: *DymoPrintQuality Text       *Resolution       300x600dpi
  *UIConstraints: *DymoPrintQuality Graphics   *Resolution       300dpi
  *UIConstraints: *Resolution       300x600dpi *DymoPrintQuality Text 
  *UIConstraints: *Resolution       300dpi     *DymoPrintQuality Graphics
```

I wonder if we should include better multi-res parsing as part of this PR.  What do you think @bberenz?  If so, do you have some printers to scrape info from?  @klabarge if you can offer some info too, that would be great.

Comment (tresf) [reply src/qz/printer/info/NativePrinter.java]: The command is:

```bash
cat "$(lpstat -l -p |grep DYMO|grep ppd|cut -d' ' -f2)" |grep dpi
```

Just replace `DYMO` with a unique printer name.

Comment (tresf) [reply src/qz/printer/info/NativePrinter.java]: My Zebra comes back with the following:

```diff
+ *DefaultResolution: 203dpi
  *Resolution 203dpi/203dpi: "<</HWResolution[203 203]/cupsBitsPerColor 1/cupsRowCount 0/cupsRowFeed 0/cupsRowStep 0/cupsColorSpace 3>>setpagedevice"
  *Resolution 300dpi/300dpi: "<</HWResolution[300 300]/cupsBitsPerColor 1/cupsRowCount 0/cupsRowFeed 0/cupsRowStep 0/cupsColorSpace 3>>setpagedevice"
  *Resolution 600dpi/600dpi: "<</HWResolution[600 600]/cupsBitsPerColor 1/cupsRowCount 0/cupsRowFeed 0/cupsRowStep 0/cupsColorSpace 3>>setpagedevice"
```

Comment (tresf) [reply src/qz/printer/info/NativePrinter.java]: My Epson WF4740 Workforce Pro shows the following (only returned one line so grep'd for `600` instead:

```diff
+ *DefaultResolution: 600x600dpi
  *cupsPrintQuality Normal: "<</cupsInteger1 4 /HWResolution[600 600]>>setpagedevice"
  *cupsPrintQuality High: "<</cupsInteger1 5 /HWResolution[600 600]>>setpagedevice"
```

Comment (klabarge) [reply src/qz/printer/info/NativePrinter.java]: > `cat "$(lpstat -l -p |grep DYMO|grep ppd|cut -d' ' -f2)" |grep dpi`

Some of the `ppd` files on my system didn't have `dpi`. I had better luck `grepping` for "Resolution".

#### Epson WF 3730
```diff
+ *DefaultResolution: 600x600dpi
*cupsPrintQuality Normal: "<</cupsInteger1 4 /HWResolution[600 600]>>setpagedevice"
*cupsPrintQuality High: "<</cupsInteger1 5 /HWResolution[600 600]>>setpagedevice"
```

#### HP LaserJet 500 Color MFP M570
This printer didn't have a `DefaultResolution` nor anything with `dpi`

```diff
/HWResolution [600 600] /PreRenderingEnhance false >> setpagedevice"
/HWResolution [600 600] /PreRenderingEnhance false >> setpagedevice"
```

#### HP LaserJet M603
This printer didn't have a `DefaultResolution`
```diff
*HPPrintQuality FastRes1200/FastRes 1200: "<</HWResolution [600 600] /PreRenderingEnhance true>> setpagedevice"
+ *HPPrintQuality 600dpi/600 dpi: "<</HWResolution [600 600] /PreRenderingEnhance false>> setpagedevice"
*HPPrintQuality ProRes1200/ProRes 1200: "<</HWResolution [1200 1200] /PreRenderingEnhance false>> setpagedevice"
```

#### Zebra LP2844
```diff
+ *DefaultResolution: 203dpi
*Resolution 203dpi/203dpi: "<</HWResolution[203 203]/cupsBitsPerColor 1/cupsRowCount 0/cupsRowFeed 0/cupsRowStep 0/cupsColorSpace 3>>setpagedevice"
*Resolution 300dpi/300dpi: "<</HWResolution[300 300]/cupsBitsPerColor 1/cupsRowCount 0/cupsRowFeed 0/cupsRowStep 0/cupsColorSpace 3>>setpagedevice"
*Resolution 600dpi/600dpi: "<</HWResolution[600 600]/cupsBitsPerColor 1/cupsRowCount 0/cupsRowFeed 0/cupsRowStep 0/cupsColorSpace 3>>setpagedevice"
```

**Comment by tresf** [comment]
> Unfortunately, due to this `TODO:` item and lack of density reporting from CUPS, we can't go higher than what's reported in the `.ppd` file as `*DefaultResolution: 300dpi` on my "Dymo LabelWriter 450 Turbo" on Mac.

I've added cross-feed parsing support to CUPS as well as additional resolution support.

@bberenz it appears you assume the first density is the best density e.g. `bestRes = rSupport.get(0);`.  Are you sorting them as you're adding them?  If so, can you help me do the same in my latest code?

**Comment by tresf** [comment]
... also @bberenz the `best` keyword isn't being honored with this new CUPS logic for some reason.

**Comment by bberenz** [comment]
>  it appears you assume the first density is the best density

No, it only grabs the first resolution to ensure not null as it then [loops through](https://github.com/qzind/tray/pull/592/files#diff-a3fee03485d53a539241fc505374f46aR158) the whole list comparing for better (or worse if draft)

**Comment by tresf** [comment]
> No, it only grabs the first resolution to ensure not null as it then loops through the whole list comparing for better (or worse if draft)

* Fixed an order-of-operations issue where the size of the densities array is zero; done by using an empty `ArrayList` to signify whether or not the densities have been fetched.  This may need eventual improvement down the road to address #607 but should fix this PR.
* Changed the logic to use `null` as default and then set on first loop instead; guards against `IndexOutOfBounds` exception in the event CUPS can't parse the densities resulting in nothing at `.get(0)`.

**Comment by tresf** [comment]
Fixed a parsing issue with `/HWResolution` where it assumed metric.  This should be ready to test.  Since many printers don't offer asymmetric densities, I'll be testing this on my Dymo Labelwriter 450, which has a density of 300x600dpi.

**Comment by tresf** [comment]
Testing is positive but also a bit inconclusive, the quality seems to remain the same between 300x300 and 600x300 on this printer, which seems rather impossible.  Perhaps the driver reports a higher resolution than what is possible, but it does not appear to hinder the output, so I'm merging.

**Pull request merged**